client/boot/index.js

```diff
@@ -369,6 +370,30 @@ function reduxStoreReady( reduxStore ) {
 		require( 'lib/rubberband-scroll-disable' )( document.body );
 	}
 
+	// 404 handler.
+	page( '*', function( context, next ) {
+		let Page404 = require( 'layout/404.jsx' );
+
+		if ( ! config.isEnabled( 'code-splitting' ) ) {
+			// If code-splitting is not enabled and we got to this route handler, no section route was matched.
+			Page404.show();
+
+			return;
+		}
+
+		var appState = context.store.getState();
```

Comment:
[Variables should be declared at the beginning of the function](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#declaring-variables-with-var).
And it's [a good case for using `const`](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#let-and-const) :)